src/polydatum/middleware.py

```diff
@@ -0,0 +1,184 @@
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta = None
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+    def __eq__(self, other):
+        return all([
+            isinstance(self, PathSegment),
+            self.name == other.name,
+            self.meta == other.meta
+        ])
+
+
+class DalMethodRequest:
+    """
+    Mutable request state. Middleware can modify this.
+    """
+
+    def __init__(
+            self, ctx: DataAccessContext, path: Tuple[PathSegment, ...], args: Tuple[Any, ...], kwargs: Dict
+    ):
+        self.ctx = ctx
+        self.path = path
+        self.args = args
+        self.kwargs = kwargs
+        self.dal_method = None  # Not resolved yet
+
+
+class DalMethodError(Exception):
+    """
+    This class defines a DalMethodError that is raised
+    when a service/method cannot be resolved by the
+    DalMethodResolverMiddleware
+    """
+    def __init__(self, path: Optional[Tuple[PathSegment]] = None):
+        self.path = path
+        super().__init__(f'Invalid dal method: {self.path}')
+
+
+class DeferredDalAttrRequest:
+    """
+    A utility class that defers access to DAL attributes.
+
+    Attribute access to the DAL is deferred so that the DAL can
+    run method middleware before and after calling a method.
+
+    Method middlewares may prevent method's from being called
+    by raising exceptions.
+    """
+
+    # Defining type here allows subclasses to easily provide another class
+    PathSegment = PathSegment
+
+    def __init__(self, handler: Callable, path: Tuple[PathSegment, ...]):
+        """
+        Args:
+            handler (Callable): A callable that handles calling the underlying
+                attribute by wrapping it so that method middleware can run before
+                and after calling a method
+            path (Tuple[PathSegment, ...]): A tuple of PathSegments representing
+                the current attribute path that has been requested.
+        """
+        self._handler = handler
+        self.path = path
+
+    def __getattr__(self, name: str) -> 'DeferredDalAttrRequest':
+        return self.__class__(self._handler, self.path + (self.PathSegment(name=name),))
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+    def __repr__(self):
+        location = [p.name for p in self.path]
+        location = '.'.join(location)
+        return f"<{self.__class__} paths={location}>"
+
+
+def dal_resolver(ctx, deferred_dal_attr_request):
+    """
+    This function resolves a dal method call to its underlying
+    service
+    """
+    service_or_method = None
+    for (_path, _service_or_method) in resolve(ctx, deferred_dal_attr_request.path):
+        if not _service_or_method:
+            raise DalMethodError(path=tuple(_path))
+        elif not isinstance(_service_or_method, (Callable, Service)):
```

Comment:
```suggestion
        if not (not _service_or_method and isinstance(_service_or_method, (Callable, Service))):
```

The result of both conditions are the same so this is a case of DRY.

Reply:
Shouldn't this be `if not _service_or_method or not isinstance(_service_or_method, (Callable, Service)):`

Reply:
I personally find a double-negative long condition harder to read.

Reply:
Sorry, I meant:

```python
if not (_service_or_method and isinstance(_service_or_method, (Callable, Service))):
```

Was only supposed to be one `not`. I read this as both items in the parens must be true together. If they are not, fail. If they were two separate `not`s, I would not as easily see the connection.

Reply:
You had me seriously questioning myself trying to work through that condition!

Reply:
I lose a lot of cred when I say "do this", but make a typo.